server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1621,22 +1619,37 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
                     request.getClusterAlias()
                 );
                 Rewriteable.rewrite(request.getRewriteable(), context, false);
+
+                if (hasRefreshPending) {
+                    final FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
+                    final MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
+                    return new CanMatchResponse(true, minMax);
+                }
+
                 final boolean aliasFilterCanMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder == false;
-                FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
-                MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
-                boolean canMatch;
-                if (canRewriteToMatchNone(request.source())) {
-                    QueryBuilder queryBuilder = request.source().query();
-                    canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
-                } else {
-                    // null query means match_all
-                    canMatch = aliasFilterCanMatch;
+                if (aliasFilterCanMatch == false
```

Comment:
Can we rewrite this as below:
1. aliasFilterCanMatch is used as double negation, better change it to aliasFilterCannotMatch
2. `sortBuilder` and `minMax` can be done in one place if `CanMatch(false, null)` is pulled up!

```
if (!hasRefreshPending) {
  final boolean aliasFilterCannotMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder;
  if (aliasFilterCannotMatch || (canRewriteToMatchNone(request.source()) && request.source().query() instanceof MatchNoneQueryBuilder)) {
    return new CanMatchResponse(false, null);
  }
}

final FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
final MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
if (hasRefreshPending) {
  return new CanMatchResponse(true, minMax);
}
```